aptos-move/framework/aptos-framework/sources/Marketplace.move

```diff
@@ -0,0 +1,289 @@
+        curr_bid: u64,
+        curr_bidder: address,
+    }
+
+    struct ListedItem has store, drop {
+        price: u64
+    }
+
+    struct CoinEscrow has key {
+        locked_coins: Table<TokenId, Coin<TestCoin>>,
+    }
+
+    struct TokenEscrow has key {
+        locked_tokens: Table<TokenId, Token>,
+    }
+
+    // Set of data sent to the event stream during a auctioning a token
+    struct AuctionEvent has drop, store {
+        id: TokenId,
+        min_bid: u64,
+        duration: u64
+    }
+
+    // Set of data sent to the event stream during a bidding for a token
+    struct BidEvent has drop, store {
+        id: TokenId,
+        bid: u64,
+    }
+
+    // Set of data sent to the event stream during a bidding for a token
+    struct ClaimCoinsEvent has drop, store {
+        id: TokenId,
+    }
+
+    // Set of data sent to the event stream during a bidding for a token
+    struct ClaimTokenEvent has drop, store {
+        id: TokenId,
+    }
+
+    // Set of data sent to the event stream during a listing of a token (for fixed price)
+    struct ListEvent has drop, store {
+        id: TokenId,
+        amount: u64,
+    }
+
+    // Set of data sent to the event stream during a buying of a token (for fixed price)
+    struct BuyEvent has drop, store {
+        id: TokenId,
+    }
+
+    struct AuctionData has key {
+        auction_items: Table<TokenId, AuctionItem>,
+        auction_events: EventHandle<AuctionEvent>,
+        bid_events: EventHandle<BidEvent>,
+        claim_coins_events: EventHandle<ClaimCoinsEvent>,
+        claim_token_events: EventHandle<ClaimTokenEvent>,
+    }
+
+    struct ListedItemsData has key {
+        listed_items: Table<TokenId, ListedItem>,
+        listing_events: EventHandle<ListEvent>,
+        buying_events: EventHandle<BuyEvent>,
+    }
+
+    public(script)fun initialize_auction(account: &signer, creator: address, collection_name: vector<u8>, name: vector<u8>, min_bid: u64, duration: u64) acquires TokenEscrow, AuctionData {
+        let token_id = Token::create_token_id_raw(creator, collection_name, name);
+        let account_addr = Signer::address_of(account);
+        if (!exists<TokenEscrow>(account_addr)) {
+            move_to(account, TokenEscrow {
+                locked_tokens: Table::new<TokenId, Token>()
+            });
+        };
+        if (!exists<AuctionData>(account_addr)) {
+            move_to(account, AuctionData {
+                auction_items: Table::new<TokenId, AuctionItem>(),
+                auction_events: Event::new_event_handle<AuctionEvent>(account),
+                bid_events: Event::new_event_handle<BidEvent>(account),
+                claim_coins_events: Event::new_event_handle<ClaimCoinsEvent>(account),
+                claim_token_events: Event::new_event_handle<ClaimTokenEvent>(account),
+            });
+        };
+        let start_time = Timestamp::now_microseconds();
+        let token = Token::withdraw_token(account, token_id, 1);
```

Comment:
This is more general comment. For auction, should we think of supporting bidding for multiple tokens in one auction?  For example, I have 3 awesome_token. I want to list them in one auction and the top 3 bidders get these 3 tokens